packages/typescript-estree/src/create-program/useProvidedProgram.ts

```diff
@@ -0,0 +1,74 @@
+import debug from 'debug';
+import * as fs from 'fs';
+import * as path from 'path';
+import * as ts from 'typescript';
+import { Extra } from '../parser-options';
+import { ASTAndProgram, getAstFromProgram } from './shared';
+
+const log = debug('typescript-eslint:typescript-estree:useProvidedProgram');
+
+function useProvidedProgram(
+  programInstance: ts.Program,
+  extra: Extra,
+): ASTAndProgram | undefined {
+  log('Retrieving ast for %s from provided program instance', extra.filePath);
+
+  programInstance.getTypeChecker(); // ensure parent pointers are set in source files
+
+  const astAndProgram = getAstFromProgram(programInstance, extra);
+
+  if (!astAndProgram) {
+    const relativeFilePath = path.relative(
+      extra.tsconfigRootDir || process.cwd(),
+      extra.filePath,
+    );
+    const errorLines = [
+      '"parserOptions.program" has been provided for @typescript-eslint/parser.',
+      `The file was not found in the provided program instance: ${relativeFilePath}`,
+    ];
+
+    throw new Error(errorLines.join('\n'));
+  }
+
+  return astAndProgram;
+}
+
+/**
+ * Utility offered by parser to help consumers construct their own program instance.
+ */
+function createProgramFromConfigFile(
+  configFile: string,
+  projectDirectory?: string,
+): ts.Program {
+  const parsed = ts.getParsedCommandLineOfConfigFile(
+    configFile,
+    { noEmit: true },
```

Comment:
I extracted a core set of options and used those. I'm really nervous about forcing users into options that could actually change the set of files that the program pulls in. Since this is just a utility for convenience, I don't want to make it unusable for folks with certain include patterns (alternatively, I don't want to encourage users to make changes to their tsconfig includes/excludes based on using this utility).